### PR TITLE
Added function to separately calculate components of dipole transition matrix

### LINKED ISSUE #33431: Crystal fields feature request: calculation of dipole transition matrix
**Is your feature request related to a problem? Please describe.**

No.

**Describe the solution you'd like**

A user (Adroja / ISIS) has request that the crystal field module be able to output the dipole transition matrix for each component (`x`, `y`, or `z` directions). At present, as a result of PR #21604 the [code](https://github.com/mantidproject/mantid/blob/9fbfecbf0cd5df1acc01bcd9fc9d2a5c97b9b54a/scripts/Inelastic/CrystalField/fitting.py#L907) can output the summed transition matrix for all components but not for each component individually. 

In addition, this feature is currently undocumented so should be included in the [interface docs](https://github.com/mantidproject/mantid/blob/9fbfecbf0cd5df1acc01bcd9fc9d2a5c97b9b54a/docs/source/interfaces/direct/Crystal%20Field%20Python%20Interface.rst).

**Describe alternatives you've considered**

The old [FOCUS](https://epubs.stfc.ac.uk/work/26115) program which the crystal field module in Mantid is meant to replaced used to output these matrices at the end of each calculation, and it is useful for users to inspect.

**Additional context**

### PR DESCRIPTION
**Pull request by MialLewis**
**Description of work.**

Separated out the `getDipoleMatrix()` into three seperate functions that calculate the individual x, y and z components of the matrix:
- `getXDipoleMatrixComponent()`
- `getYDipoleMatrixComponent()`
- `getZDipoleMatrixComponent()`

Added unit test to ensure in the future the sum of the three component functions equals the return value of the `getDipoleMatrix()` function.

**To test:**

1) Create crystal field object (e.g `cf = CrystalField.CrystalField('Ce', 'C2v', B20=0.035, B40=-0.012, B43=-0.027, B60=-0.00012, B63=0.0025, B66=0.0068)`)
2) call getDipoleMatrix() function of object and observe that is returns a matrix that is equal to that output by previous versions of mantid
3) confirm `test_matrix_components_equal_to_summed_matrix`  unit test passed asserting equivalence of the sum of component matrix and return value of getDipoleMatrix() function 

Resolves #33431 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by cailafinn — CHANGES_REQUESTED** [review]
Works well, just a couple of minor things.

As a note on some git stuff: 
Our [standards](https://developer.mantidproject.org/GitWorkflow.html#naming-branches) suggest that banches should be named with the issue number first and then a very short description of what the issue is. 
It can also be worth including a ref at the end of the commit message to the issue, which can be useful when looking at the blame to see why a change was made. E.g `Fix thing RE #33573`

File: scripts/Inelastic/CrystalField/fitting.py [hunk 1]
```diff
@@ -904,19 +904,40 @@ def getMagneticMoment(self, *args, **kwargs):
...
-        _, _, hy = energies(self._nre, BextY=1.0)
-        _, _, hz = energies(self._nre, BextZ=1.0)
         ix = np.dot(np.conj(np.transpose(self._eigenvectors)), np.dot(hx, self._eigenvectors))
+        return np.multiply(ix, np.conj(ix))/(gJuB ** 2)
+
+    def getYDipoleMatrixComponent(self, gJuB = None):
+        self._calcEigensystem() #will not recalculate if already called (unless _dirty_eigensystem)
+        if gJuB is None:
+            gJuB = self._calc_gJuB()
+        _, _, hy = energies(self._nre, BextY=1.0)
```
[the remark this answers is not in the document]
Comment (MialLewis) [reply scripts/Inelastic/CrystalField/fitting.py:924]: I did toy with having one `getDipoleMatrixComponent` function that takes a string (x, y or z) as an argument. I wasn't sure which I preferred - the simplicity of having three self explanatory functions with no argument requirements/need to refer to docs, or slightly reduced code with an argument required, albeit a simple one, and the associated error handling etc.

What do you think?

**Comment by MialLewis** [comment]
> Works well, just a couple of minor things.
> 
> As a note on some git stuff: Our [standards](https://developer.mantidproject.org/GitWorkflow.html#naming-branches) suggest that banches should be named with the issue number first and then a very short description of what the issue is. It can also be worth including a ref at the end of the commit message to the issue, which can be useful when looking at the blame to see why a change was made. E.g `Fix thing RE #33573`

Noted

File: scripts/Inelastic/CrystalField/fitting.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MialLewis) [reply scripts/Inelastic/CrystalField/fitting.py]: We've decided to consolidate into one function.

File: scripts/Inelastic/CrystalField/fitting.py [hunk 2]
```diff
@@ -905,38 +905,32 @@ def getMagneticMoment(self, *args, **kwargs):
...
-        _, _, hy = energies(self._nre, BextY=1.0)
-        iy = np.dot(np.conj(np.transpose(self._eigenvectors)), np.dot(hy, self._eigenvectors))
-        return np.multiply(iy, np.conj(iy))/(gJuB ** 2)
+        if nComponent == 'X' or nComponent == 'x':
+            _, _, h_n = energies(self._nre, BextX=1.0)
+        elif nComponent == 'Y' or nComponent == 'y':
+            _, _, h_n = energies(self._nre, BextY=1.0)
+        elif nComponent == 'Z' or nComponent == 'z':
+            _, _, h_n = energies(self._nre, BextZ=1.0)
+        else:
```
Comment (ConorMFinn) [inline scripts/Inelastic/CrystalField/fitting.py:923]: This if block could be simplified by calling `nComponent = nComponent.lower()` or `.upper()` before checking them

**Review by cailafinn — CHANGES_REQUESTED** [review]
This looks good, just one more minor thing:

File: scripts/Inelastic/CrystalField/fitting.py (hunk 2 above)
Comment (MialLewis) [reply scripts/Inelastic/CrystalField/fitting.py:923]: I did consider this. but wanted my error message detailing what the inputs should be to be thrown if a non string be passed in for whatever reason.

Comment (ConorMFinn) [reply scripts/Inelastic/CrystalField/fitting.py:923]: That can still be there, it would just mean that you wouldn't need the `or` elements in each check

Comment (MialLewis) [reply scripts/Inelastic/CrystalField/fitting.py:923]: would int.lower() not throw an unhandled exception?

Comment (ConorMFinn) [reply scripts/Inelastic/CrystalField/fitting.py:923]: `nComponent` is a string, isn't it?

Comment (MialLewis) [reply scripts/Inelastic/CrystalField/fitting.py:923]: my thoughts exactly, probably not much in it!

**Review by cailafinn — APPROVED** [review]
Works well, nice work 👍

**Pull request merged**